proof/refine/ARM/StateRelation.thy

```diff
@@ -281,9 +281,29 @@ lemma refills_heads_equal_valid_sched_context':
    \<Longrightarrow> rAmount (refillHd sc') = r_amount (refill_hd sc)
        \<and> rTime (refillHd sc') = r_time (refill_hd sc)"
   apply (frule refills_heads_equal; (solves simp)?)
-  apply (auto simp: sc_relation_def valid_sched_context'_def refillHd_def refills_map_def
-                    refill_map_def wrap_slice_def
-             split: if_splits)
+  by (auto simp: sc_relation_def valid_sched_context'_def refillHd_def refills_map_def
+                 refill_map_def wrap_slice_def
+          split: if_splits)
+
+(* Most sched contexts should satisfy these conditions. The lemma below (valid_refills'_common_usage)
+   illustrates a common situation where these conditions hold.
+
+   It seems clear that an abbreviation of this kind is useful, but it's not clear exactly which
+   conditions it should contain. For instance, it does not contain "0 < scRefillMax sc" which one
+   may argue is equally standard.
+*)
+abbreviation (input) valid_refills' where
+  "valid_refills' sc \<equiv> scRefillMax sc \<le> length (scRefills sc) \<and> scRefillHead sc < scRefillMax sc \<and>
+                    scRefillCount sc \<le> scRefillMax sc \<and> 0 < scRefillCount sc"
```

Comment:
With its new name and all that, I would be happy to have this abbreviation without `(input)`. I have started to use this with projections and it seems to make sense. We can have this as a conjunct in `valid_sched_context’`. Not quite sure if we should handle this separately yet, although I realise we do have `valid_refills` as a notion separate from `valid_obj`.

Reply:
Looking at the merged files, the previous version of this (`std_sc’)` is still sitting towards the end of this file. You originally called this one here as `standard_sc` and that made me think it was the same thing as your `std_sc’` from the previous PR.  But does this mean there was actually some duplication? Surely we don’t need both, do we?

Reply:
I can clean it up in my PR.

Reply:
Any duplication is a mistake, I was only intending to use one notion.